source/common/common/compiled_string_map.h

```diff
@@ -0,0 +1,208 @@
+        return {};
+      }
+      return branches_[k - min_]->find(key);
+    }
+
+  private:
+    const size_t index_;
+    const uint8_t min_;
+    const std::vector<std::unique_ptr<Node>> branches_;
+  };
+
+public:
+  using KV = std::pair<absl::string_view, Value>;
+  /**
+   * Returns the value with a matching key, or the default value
+   * (typically nullptr) if the key was not present.
+   * @param key the key to look up.
+   */
+  Value find(absl::string_view key) const {
+    const size_t key_size = key.size();
+    if (key_size >= table_.size() || table_[key_size] == nullptr) {
+      return {};
+    }
+    return table_[key_size]->find(key);
+  };
+  /**
+   * Construct the lookup table. This can be a somewhat slow multi-pass
+   * operation if the input table is large.
+   * @param initial a vector of key->value pairs. This is taken by value because
+   *                we're going to modify it. If the caller still wants the original
+   *                then it can be copied in, if not it can be moved in.
+   *                Note that the keys are string_views - the base string data must
+   *                exist for the duration of compile(). The leaf nodes take copies
+   *                of the key strings, so the string_views can be invalidated once
+   *                compile has completed.
+   */
+  void compile(std::vector<KV> initial) {
+    if (initial.empty()) {
+      return;
+    }
+    std::sort(initial.begin(), initial.end(),
+              [](const KV& a, const KV& b) { return a.first.size() < b.first.size(); });
+    size_t longest = initial.back().first.size();
+    table_.resize(longest + 1);
+    auto range_start = initial.begin();
+    // Populate the sub-nodes for each length of key that exists.
+    while (range_start != initial.end()) {
+      // Find the first key whose length differs from the current key length.
+      // Everything in between is keys with the same length.
+      auto range_end =
+          std::find_if(range_start, initial.end(), [len = range_start->first.size()](const KV& e) {
+            return e.first.size() != len;
+          });
+      std::vector<KV> node_contents;
+      // Populate a FindFn for the nodes in that range.
+      node_contents.reserve(range_end - range_start);
+      std::copy(range_start, range_end, std::back_inserter(node_contents));
+      table_[range_start->first.size()] = createEqualLengthNode(node_contents);
+      range_start = range_end;
+    }
+  }
+
+private:
+  /*
+   * @param node_contents the set of key-value pairs that will be children of
+   *                      this node.
+   */
+  static std::unique_ptr<Node> createEqualLengthNode(std::vector<KV> node_contents) {
+    if (node_contents.size() == 1) {
+      return std::make_unique<LeafNode>(node_contents[0].first, std::move(node_contents[0].second));
+    }
+    struct IndexSplitInfo {
+      uint8_t index, min, max, count;
```

Comment:
envoy style uses _ to end struct member vars (google style omits the _ for struct0